### PR TITLE
Lazy evaluation

### PR DESCRIPTION
**Pull request by vonglasow**
Add lazy operator

### CODE REVIEW

File: Model/Operator.php [hunk 1]
```diff
@@ -52,6 +52,21 @@
...
 
     /**
+     * Break if operator is Lazy and the evaluation is false.
+     *
+     * @const string
+     */
+    const LAZY_BREAK = false;
+
+    /**
+     * Continue if operator is Lazy and the evaluation is true.
```
Comment (Hywan) [inline Model/Operator.php]: “and if the value does not break the evaluation”

File: Model/Operator.php [hunk 2]
```diff
@@ -52,6 +52,21 @@
 class Operator implements Visitor\Element {
 
     /**
+     * Break if operator is Lazy and the evaluation is false.
```
Comment (Hywan) [inline Model/Operator.php]: “and if the current argument break the evaluation”

File: Model/Operator.php [hunk 3]
```diff
@@ -72,7 +87,12 @@ class Operator implements Visitor\Element {
      */
     protected $_function  = true;
 
-
+    /**
+     * Use this property to define lazy operator or not
```
Comment (Hywan) [inline Model/Operator.php]: “Use this property to define whether the operator is lazy or not.”

File: Model/Operator.php [hunk 4]
```diff
@@ -178,6 +199,46 @@ public function isFunction ( ) {
...
+     * @return  bool
+     */
+    public function isLazy() {
+
+        return $this->_lazy;
+    }
+
+    public function lazyEvaluate ( $value )
+    {
+        if (!$this->isLazy()) return self::LAZY_CONTINUE;
```
Comment (Hywan) [inline Model/Operator.php]: Good :+1:!

File: Model/Operator.php [hunk 5]
```diff
@@ -178,6 +199,46 @@ public function isFunction ( ) {
...
+
+        switch ($this->_name) {
+            case 'and':
+                return $value === self::LAZY_BREAK;
```
Comment (Hywan) [inline Model/Operator.php]: If we change the constant values, it will break this code. I think we should go for something more explicit:
```php
switch($this->_name) {
    case 'and':
        if(false === $value)
            return self::LAZY_BREAK;
      break;

    case 'or':
        if(true === $value)
            return self::LAZY_BREAK;
     break;
}

return self::LAZY_CONTINUE;
```

I think it's better to define only the `LAZY_BREAK` cases. All the other are `LAZY_CONTINUE` by definition.

File: Visitor/Asserter.php [hunk 6]
```diff
@@ -154,13 +154,20 @@ protected function visitOperator ( Visitor\Element $element, &$handle = null, $e
...
+
+            $arguments[] = $value;
+        }
 
         if(false === $this->operatorExists($name))
             throw new Ruler\Exception\Asserter(
                 'Operator %s does not exist.', 1, $name);
 
+        if ( 2 > count($arguments) ) return null;
+
```
Comment (Hywan) [inline Visitor/Asserter.php]: Why?

Comment (vonglasow) [inline Visitor/Asserter.php]: Because, if we break the loop we still passe arguments in the next step which create an error because arguments should be an array with at least 2 elements. And 3 of 4 tests failed, 

Errors raised are
```sh
=> Hoa\Ruler\Test\Unit\Documentation::case_classical():
==> Error E_NOTICE in /home/vagrant/ruler/Test/Unit/Documentation.php on line 62, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Undefined variable: b
==> Error E_WARNING in /home/vagrant/ruler/Test/Unit/Documentation.php on line 62, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Missing argument 2 for Hoa\Ruler\Visitor\Asserter::Hoa\Ruler\Visitor\{closure}()
=> Hoa\Ruler\Test\Unit\Documentation::case_new_operators():
==> Error E_NOTICE in /home/vagrant/ruler/Test/Unit/Documentation.php on line 121, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Undefined variable: a
==> Error E_WARNING in /home/vagrant/ruler/Test/Unit/Documentation.php on line 121, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Missing argument 2 for Hoa\Ruler\Visitor\Asserter::Hoa\Ruler\Visitor\{closure}()
==> Error E_WARNING in /home/vagrant/ruler/Test/Unit/Documentation.php on line 121, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Missing argument 1 for Hoa\Ruler\Visitor\Asserter::Hoa\Ruler\Visitor\{closure}()
=> Hoa\Ruler\Test\Unit\Documentation::case_interprete():
==> Error E_NOTICE in /home/vagrant/ruler/Test/Unit/Documentation.php on line 138, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Undefined variable: b
==> Error E_WARNING in /home/vagrant/ruler/Test/Unit/Documentation.php on line 138, generated by file /home/vagrant/ruler/Visitor/Asserter.php on line 84:
Missing argument 2 for Hoa\Ruler\Visitor\Asserter::Hoa\Ruler\Visitor\{closure}()
```

Comment (Hywan) [inline Visitor/Asserter.php]: Yes of course… I don't know if this is the good fix. I will think about it.

Comment (vonglasow) [inline Visitor/Asserter.php]: I am sure it's not a good fix but it was the only one I have found for now.

File: Model/Operator.php [hunk 7]
```diff
@@ -178,6 +200,48 @@ public function isFunction ( ) {
...
+     * @access  protected
+     * @param   bool  $isLazy    Is a lazy or not.
+     * @return  bool
+     */
+    protected function setLazy ( $isLazy ) {
+
+        $old         = $this->_lazy;
+        $this->_lazy = (bool) $isLazy;
+
+        return $old;
```
Comment (stephpy) [inline Model/Operator.php]: Why is it useful to return old value of `$this->_lazy` ?

Comment (Hywan) [inline Model/Operator.php]: Because this is how setter works in Hoa since the beginning :-).

Comment (stephpy) [inline Model/Operator.php]: Ha, so the discussion about usefulness of it isn't here ? :p 

In this example, we don't care about old value.

Comment (Hywan) [inline Model/Operator.php]: ;-)

**Comment by Hywan** [comment]
Good :-). Is it finished?

**Comment by vonglasow** [comment]
I thinks it looks finish I will try again if it works well and confirm it!

**Comment by vonglasow** [comment]
Test added, if test ok and nothing missing I think is done.

**Comment by Hywan** [comment]
Please, `git rebase master` to ease the review :-).

File: Model/Operator.php [hunk 8]
```diff
@@ -52,6 +52,22 @@
 class Operator implements Visitor\Element {
 
     /**
+     * Break if operator is Lazy and if the current argument value does not
```
Comment (Hywan) [inline Model/Operator.php]: Why a capital `L` to `Lazy`?

File: Model/Operator.php [hunk 9]
```diff
@@ -52,6 +52,22 @@
...
+     * break the evaluation.
+     *
+     * @const string
+     */
+    const LAZY_BREAK = false;
+
+    /**
+     * Continue if operator is Lazy and if the current argument value breaks
```
Comment (Hywan) [inline Model/Operator.php]: Same here.

File: Model/Operator.php [hunk 10]
```diff
@@ -87,6 +108,7 @@ public function __construct ( $name, Array $arguments = [],
                                   $isFunction = true ) {
 
         $this->setName($name);
+        $this->setLazy($name === 'and' || $name === 'or');
```
Comment (Hywan) [inline Model/Operator.php]: We used to move the most constant part to the left of operators, thus:
```
$this->setLazy('and' === $name || 'or' === $name)
```

File: Model/Operator.php [hunk 11]
```diff
@@ -178,6 +200,48 @@ public function isFunction ( ) {
...
+     * Set whether the operator is lazy or not.
+     *
+     * @access  protected
+     * @param   bool  $isLazy    Is a lazy or not.
+     * @return  bool
+     */
+    protected function setLazy ( $isLazy ) {
+
+        $old         = $this->_lazy;
+        $this->_lazy = (bool) $isLazy;
```
Comment (Hywan) [inline Model/Operator.php]: No need to cast to `bool`. We expect a bool, developer must respect the contract :-).

File: Model/Operator.php [hunk 12]
```diff
@@ -178,6 +200,48 @@ public function isFunction ( ) {
...
+     *
+     * @access  public
+     * @return  bool
+     */
+    public function isLazy() {
+
+        return $this->_lazy;
+    }
+
+    public function lazyEvaluate ( $value )
```
Comment (Hywan) [inline Model/Operator.php]: We must find a better name for this method. Also, its API document is missing :-).

File: Test/Unit/Documentation.php [hunk 13]
```diff
@@ -172,4 +172,35 @@ public function case_compile ( ) {
                 ->string($result)
                     ->isEqualTo($expectedResult);
     }
+
+    public function case_lazyOperator()
```
Comment (Hywan) [inline Test/Unit/Documentation.php]: This must be written in another test suite (another file), for example: `Test/Unit/Visitor/Asserter.php`.

File: Test/Unit/Documentation.php [hunk 14]
```diff
@@ -172,4 +172,35 @@ public function case_compile ( ) {
...
+    {
+        $this
+            ->given(
+                $ruler = new LUT(),
+                $fUse  = false,
```
Comment (Hywan) [inline Test/Unit/Documentation.php]: `f` and `g` are used but not necesarily executed. Maybe you should rename these variables into `$fActivated` or `$fExecuted`, same for `$g…`.

File: Test/Unit/Documentation.php [hunk 15]
```diff
@@ -172,4 +172,35 @@ public function case_compile ( ) {
...
+                    return $a;
+                }),
+                $ruler->getDefaultAsserter()->setOperator('g', function ( $b = false ) use ( &$gUse ) {
+                    $gUse = true;
+                    return $b;
+                }),
+                $rule             = 'f(false) and g(true)',
+                $context          = new LUT\Context(),
+                $context['true']  = true,
+                $context['false'] = false
```
Comment (Hywan) [inline Test/Unit/Documentation.php]: We don't have the `trust` and `false` constants?

File: Test/Unit/Documentation.php [hunk 16]
```diff
@@ -172,4 +172,35 @@ public function case_compile ( ) {
...
+            )
+            ->when($result = $ruler->assert($rule, $context))
+            ->then
+                ->boolean($result)
+                ->isFalse()
+                ->boolean($fUse)
+                ->isTrue()
+                ->boolean($gUse)
+                ->isFalse()
+            ;
```
Comment (Hywan) [inline Test/Unit/Documentation.php]: We have a matrix to test:
  1. `f(true) and g(true)`,
  2. `f(true) and g(false)`,
  3. `f(false) and g(true)`,
  4. `f(false) and g(false)`.

If we only consider the lazyness execution, yup, only 1 and 3 are useful but we want to be sure we didn't break something, so we need to test all these test cases.

File: Visitor/Asserter.php [hunk 17]
```diff
@@ -81,8 +81,8 @@ public function __construct ( Ruler\Context $context = null ) {
         if(null !== $context)
             $this->setContext($context);
 
-        $this->setOperator('and', function ( $a, $b ) { return $a && $b; });
-        $this->setOperator('or',  function ( $a, $b ) { return $a || $b; });
+        $this->setOperator('and', function ( $a = false, $b  = false ) { return $a && $b; });
+        $this->setOperator('or',  function ( $a = false, $b = false ) { return $a || $b; });
         $this->setOperator('xor', function ( $a, $b ) { return (bool) ($a ^ $b); });
```
Comment (Hywan) [inline Visitor/Asserter.php]: Could we add lazyness for this operator too?

Comment (vonglasow) [inline Visitor/Asserter.php]: I don't know, for me, xor it's only if the first or the second arguments of the operator is true, and never both. So we always need to evaluate the second argument regardless of the result of the first evaluation.

File: Test/Unit/Documentation.php (hunk 13 above)
Comment (vonglasow) [inline Test/Unit/Documentation.php]: ok ;)

File: Visitor/Asserter.php (hunk 17 above)
Comment (Hywan) [inline Visitor/Asserter.php]: Exact ;-).

File: Test/Unit/Documentation.php (hunk 13 above)
Comment (Hywan) [inline Test/Unit/Documentation.php]: Either we put it in `Test/Unit/Model/Operator.php` or in `Test/Unit/Visitor/Asserter.php`, but this is something introduced in the model, so you're right, we have to put it in `Test/Unit/Model/Operator.php` :+1:.

File: Test/Unit/Model/Operator.php [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+                ->boolean($result)
+                    ->isFalse()
+                ->boolean($fExecuted)
+                    ->isTrue()
+                ->boolean($gExecuted)
+                    ->isTrue()
+            ;
+    }
+
+    public function case_xor()
```
Comment (Hywan) [inline Test/Unit/Model/Operator.php]: This test case is not about lazy evaluation, so it is not part of this PR. Could you remove it and keep it in another branch (you could make a new PR with this one)?

File: Model/Operator.php [hunk 19]
```diff
@@ -178,6 +200,56 @@ public function isFunction ( ) {
...
+     * @access  public
+     * @return  bool
+     */
+    public function isLazy() {
+
+        return $this->_laziness;
+    }
+
+    /**
+     * Evaluate the lazyness of operator to know if we break or continue.
```
Comment (Hywan) [inline Model/Operator.php]: Proposal: `Check whether we should break the lazy evaluation or not.`

File: Test/Unit/Model/Operator.php [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+use Hoa\Test;
+use Hoa\Ruler as LUT;
+
+/**
+ * Class \Hoa\Ruler\Test\Unit\Documentation.
+ *
+ * Test suite of the examples in the documentation.
+ *
+ * @author     Hoa community <avonglasow@hoa-project.net>
```
Comment (Hywan) [inline Test/Unit/Model/Operator.php]: Update the author.

File: Test/Unit/Model/Operator.php [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+ * @copyright  Copyright © 2007-2015 Hoa community.
```
Comment (Hywan) [inline Test/Unit/Model/Operator.php]: Also here.

**Comment by Hywan** [comment]
Congrats :-).

**Comment by stephpy** [comment]
:+1:
